### PR TITLE
add history items to elation actions

### PR DESCRIPTION
**Pull request by sharlotta93**
### **PR Type**
Enhancement, Tests

___

### **Description**
- Implemented the `addHistories` action to allow adding history items to Elation's patient page.
- Added comprehensive tests for the `addHistories` action, ensuring successful completion and error handling.
- Defined new constants and mock functions for history items in the client and constants files.
- Updated the Elation client to include methods for handling history items.
- Created new types and validation schemas for history items.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>9 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>client.ts</strong><dd><code>Add mock function for history response in client</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/__mocks__/client.ts

<li>Added <code>historyResponseExample</code> to the import list.<br> <li> Implemented <code>addHistory</code> mock function returning <code>historyResponseExample</code>.<br> <br>

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/507/files#diff-f233b91df2b9bf19f5f13ccc7522c18dd4a8085ebdb2812dc8dfab8a04372049">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>constants.ts</strong><dd><code>Define history example constants for testing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/__mocks__/constants.ts

- Added `historyExample` and `historyResponseExample` constants.

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/507/files#diff-1620308b774afef7cdd55c47a2fc1c5d728f51202540645730dc217894556fb6">+17/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>addHistories.ts</strong><dd><code>Implement addHistories action with error handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/actions/addHistories/addHistories.ts

<li>Implemented <code>addHistories</code> action for adding history items.<br> <li> Included error handling for API and validation errors.<br>

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/507/files#diff-23a49011f8fba635f601bd8234f83c227c43fef1d4767069de4f1103b104b652">+91/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>fields.ts</strong><dd><code>Define fields and validation for history items</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/actions/addHistories/config/fields.ts

- Defined fields for history items with validation schema.

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/507/files#diff-4162c845fdde68b53447a594d001b3f4323893b94522935f684619c37275df68">+124/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Export fields and validation schema</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/actions/addHistories/config/index.ts

- Exported fields and validation schema from fields.ts.

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/507/files#diff-1435cbb00d055a8e664592bfa05c170fb1722d30c545daa72538a7d8dfc9ae4e">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Export addHistories action</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/actions/addHistories/index.ts

- Exported `addHistories` action.

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/507/files#diff-4891f1a75f26b26f08bb986c24aa9649ce30dfba21b16595487a19b8b4f70028">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Include addHistories in exported actions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/actions/index.ts

- Added `addHistories` to the list of exported actions.

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/507/files#diff-9a1c0055b212081498d2441b821c3c51501a8ff23fa020ebccd3e99395c37a08">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>client.ts</strong><dd><code>Add addHistory method to Elation client</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/client.ts

<li>Added <code>addHistory</code> method to <code>ElationDataWrapper</code> and <code>ElationAPIClient</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/507/files#diff-65c140fbe1debe73b11455b75ab0aa6aa573b3c47babb49245c0d35c8e62a662">+15/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>history.ts</strong><dd><code>Define types and schemas for history items</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/types/history.ts

- Defined types and schemas for history items.

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/507/files#diff-3a2929a1ed1c77a0f24e39a72fc7fa98a9e64840ac11830a0b56f7fd67c3d8a2">+39/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>addHistories.test.ts</strong><dd><code>Add tests for addHistories action</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

extensions/elation/actions/addHistories/addHistories.test.ts

<li>Created tests for <code>addHistories</code> action.<br> <li> Tested successful completion and error handling.<br>

</details>

  </td>
  <td><a href="https://github.com/awell-health/awell-extensions/pull/507/files#diff-7b5b7694b181d21c98147280908a35d2164aba92e4cd1b9c6e48e75fe606e211">+63/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: extensions/elation/types/history.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    type: HistoryTypeSchema,
+    patient: NumericIdSchema,
+    text: z.string(),
+  })
+  .strict()
+
+export const historyArraySchema = z.array(historySchema)
+
+export type addHistoryInput = z.infer<typeof historySchema>
+export interface addHistoryResponse extends addHistoryInput {
```
Comment (bejoinka) [inline extensions/elation/types/history.ts]: Github is doing my work for me

File: extensions/elation/actions/addHistories/addHistories.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+              },
+              error: {
+                category: 'SERVER_ERROR',
+                message: `${err.status ?? '(no status code)'} Error: ${
+                  err.message
+                }`,
+              },
+            },
+          ],
+        })
```
Comment (bejoinka) [inline extensions/elation/actions/addHistories/addHistories.ts:84]: we handle axios errors in the server as well

File: extensions/elation/actions/addHistories/addHistories.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            failedFields.push(`${type}: ${errMessage}`)
+          }
+        })
+      )
+
+      if (failedFields.length > 0) {
+        throw new Error(
+          `Failed to create histories for fields: ${failedFields.join(', ')}`
+        )
+      }
```
Comment (bejoinka) [inline extensions/elation/actions/addHistories/addHistories.ts:65]: there's a bit of a problem here. if 9/10 fields pass and one fails, we'll still be writing to the histories for 9/10 of them.
so, i think throwing an error here is not quite right.

do you know of a way we can guarantee all of the addHistory calls happen in a transaction?

**Review by bejoinka — CHANGES_REQUESTED** [review]
i've left some requests for changes. i see a problem in the way the action executes. LMK how you'd like to handle

File: extensions/elation/types/history.ts (hunk 1 above)
Comment (sharlotta93) [reply extensions/elation/types/history.ts]: that's fixed

File: extensions/elation/actions/addHistories/addHistories.ts (hunk 3 above)
Comment (sharlotta93) [reply extensions/elation/actions/addHistories/addHistories.ts:65]: I did look into that - the API currently doesn't support adding multiple histories so we would have to manually delete them 1 by 1 (which can also throw errors in which case we end up in ugly mid-state).  I think this is a bit of a corner case and I've added this as a way to let the user know - hey this is missing. So I think for v1 that should be enough. LMK if you disagree.

Comment (sharlotta93) [reply extensions/elation/actions/addHistories/addHistories.ts]: @bejoinka let me know if I'm okay to merge or if you want to chat more about the ^

Comment (bejoinka) [reply extensions/elation/actions/addHistories/addHistories.ts]: yeah, i had made this one open ended but indeed i don't think it's a good idea to "fail" an activity that successfully writes to the API. Is the expectation here that people will write one history at a time?

**Pull request merged**